puffin/src/lib.rs

```diff
@@ -231,24 +231,47 @@ macro_rules! profile_function {
     };
 }
 
-#[allow(clippy::doc_markdown)] // clippy wants to put "MacBook" in ticks 🙄
+
 /// Profile the current scope with the given name (unique in the parent scope).
+/// 
+/// This macro is identical to [profile_scope], except that it expands to the expression
+/// containing the profiling scope, as opposed to [profile_scope] which expands to a
+/// variable (which cannot be accessed due to macro hygiene).
 ///
-/// Names should be descriptive, ASCII and without spaces.
-///
-/// Example: `profile_scope!("load_mesh");`.
+/// This allows for profiling scopes to persist for a custom duration.
+/// 
+/// # Example
+/// 
+/// ```rust
+/// # use std::iter::FromIterator as _;
+/// # 
+/// # pub mod rayon { pub mod prelude {
+/// #     pub fn for_each_init<T, I>(vec: &std::vec::Vec<T>, init: fn() -> I, body: fn ((I, T)) -> ()) {
+/// #     }
+/// # } }
+/// #
+/// let some_large_vec = Vec::from_iter(0..1000);
 ///
-/// An optional second argument can be a string (e.g. a mesh name) to help diagnose what was slow.
-/// Example: `profile_scope!("load_mesh", mesh_name);`
-///
-/// Overhead: around 54 ns on Macbook Pro with Apple M1 Max.
+/// // Use rayon's parallel for loop over our large iterator
+/// rayon::prelude::for_each_init(
+///         &some_large_vec,
+///         // This gets called to init each work segment, and is passed into the calls
+///         // Rayon keeps the profiling scope stored for the entire duration of the work segment
+///         // So we can track the entire segment as one, instead of each loop iteration
+///         || puffin::profile_scope_custom!("rayon_work_segment"),
+///         |((_profiler_scope), i)| {
+///             // All calls here gets profiled on the same scope
+///             println!("{i}")
+///         },
+/// );
+/// ```
 #[macro_export]
-macro_rules! profile_scope {
+macro_rules! profile_scope_custom {
     ($name:expr) => {
-        $crate::profile_scope!($name, "");
+        $crate::profile_scope_custom!($name, "")
     };
     ($name:expr, $data:expr) => {
```

Comment:
These should be double-brackets to make the macro usable in more scenarios.

Reply:
That would mean it expands to a `BlockExpression` not `Expression` ([ref](https://doc.rust-lang.org/reference/expressions.html)).

What sort of scenarios would benefit from this? Could you perhaps list an example where the adding second brackets would be required?

Reply:
Take code like this:

```
match whatever {
    Ok(x) => my_macro!(),
    _ => (),
}
```

A multi-line statement cannot be present in this location, which is why it's generally good practice to wrap the macro in an extra set of brackets (which you'll see for many Rust macros like [println](https://doc.rust-lang.org/src/std/macros.rs.html#138-145)).

For this specific macro it might be fine, since the entire macro body is already a block due to the conditional, but this is extremely easy to break by just adding one more line outside of it.

Reply:
Done